pkg/sql/distsqlrun/algebraic_set_op.go

```diff
@@ -88,183 +98,250 @@ func newAlgebraicSetOp(
 }
 
 func (e *algebraicSetOp) Run(wg *sync.WaitGroup) {
+	if e.out.output == nil {
+		panic("algebraicSetOp output not initialized for emitting rows")
+	}
+	Run(e.flowCtx.Ctx, e, e.out.output)
 	if wg != nil {
-		defer wg.Done()
+		wg.Done()
 	}
+}
 
-	ctx := log.WithLogTag(e.flowCtx.Ctx, "ExceptAll", nil)
-	ctx, span := processorSpan(ctx, "exceptAll")
-	defer tracing.FinishSpan(span)
-
-	log.VEventf(ctx, 2, "starting exceptAll set process")
-	defer log.VEventf(ctx, 2, "exiting exceptAll")
+// producerMeta constructs the ProducerMetadata after consumption of rows has
+// terminated, either due to being indicated by the consumer, or because the
+// processor ran out of rows or encountered an error. It is ok for err to be
+// nil indicating that we're done producing rows even though no error occurred.
+func (e *algebraicSetOp) producerMeta(err error) *ProducerMetadata {
+	var meta *ProducerMetadata
+	if !e.closed {
+		if err != nil {
+			meta = &ProducerMetadata{Err: err}
+		} else if trace := getTraceData(e.ctx); trace != nil {
+			meta = &ProducerMetadata{TraceData: trace}
+		}
+		// We need to close as soon as we send producer metadata as we're done
+		// sending rows. The consumer is allowed to not call ConsumerDone().
+		e.internalClose()
+	}
+	return meta
+}
 
-	defer e.leftSource.ConsumerDone()
-	defer e.rightSource.ConsumerDone()
+func (e *algebraicSetOp) Next() (sqlbase.EncDatumRow, *ProducerMetadata) {
+	var row sqlbase.EncDatumRow
+	var meta *ProducerMetadata
 
-	e.evalCtx = e.flowCtx.NewEvalCtx()
+	for {
+		switch e.opType {
+		case AlgebraicSetOpSpec_Except_all:
+			row, meta = e.nextExceptAll()
+		default:
+			panic(fmt.Sprintf("cannot run unsupported algebraicSetOp %v", e.opType))
+		}
 
-	switch e.opType {
-	case AlgebraicSetOpSpec_Except_all:
-		err := e.exceptAll(ctx)
-		DrainAndClose(ctx, e.out.output, err, e.leftSource, e.rightSource)
+		if e.closed || meta != nil {
+			return nil, meta
+		}
+		if row == nil {
+			return nil, e.producerMeta(nil /* err */)
+		}
 
-	default:
-		panic(fmt.Sprintf("cannot run unsupported algebraicSetOp %v", e.opType))
+		outRow, status, err := e.out.ProcessRow(e.ctx, row)
+		if err != nil {
+			return nil, e.producerMeta(err)
+		}
+		switch status {
+		case NeedMoreRows:
+			if outRow == nil && err == nil {
+				log.Infof(e.evalCtx.Ctx(), "outrow is nil, but need more rows. skipping")
```

Comment:
Not familiar with logging in Distsql land, but I would suggest putting this behind a V(2) level log if this is just an ordinary Infof.